src/write_resume_data.cpp

```diff
@@ -285,6 +296,88 @@ namespace libtorrent {
 		return ret;
 	}
 
+	entry write_torrent_file(add_torrent_params const& atp)
+	{
+		entry ret;
+		if (!atp.ti)
+			aux::throw_ex<system_error>(errors::torrent_missing_info);
+
+		auto const info = atp.ti->info_section();
+		ret["info"].preformatted().assign(info.data(), info.data() + info.size());
+		if (!atp.ti->comment().empty())
+			ret["comment"] = atp.ti->comment();
+		if (atp.ti->creation_date() != 0)
+			ret["creation date"] = atp.ti->creation_date();
+		if (!atp.ti->creator().empty())
+			ret["created by"] = atp.ti->creator();
+
+		if (!atp.merkle_trees.empty())
+		{
+			file_storage const& fs = atp.ti->files();
+			auto& trees = atp.merkle_trees;
+			if (int(trees.size()) != fs.num_files())
+				aux::throw_ex<system_error>(errors::torrent_missing_piece_layer);
+
+			auto& piece_layers = ret["piece layers"].dict();
+			for (file_index_t f : fs.file_range())
+			{
+				if (fs.pad_file_at(f) || fs.file_size(f) < fs.piece_length())
+					continue;
+
+				aux::merkle_tree t(fs.file_num_blocks(f)
+					, fs.piece_length() / default_block_size, fs.root_ptr(f));
+
+				auto const& tree = trees[f];
+				if (f < atp.merkle_tree_mask.end_index() && !atp.merkle_tree_mask[f].empty())
+				{
+					t.load_sparse_tree(tree, atp.merkle_tree_mask[f]);
+				}
+				else
+				{
+					t.load_tree(tree);
+				}
+
+				auto const piece_layer = t.get_piece_layer();
+				if (int(piece_layer.size()) != fs.file_num_pieces(f))
+					aux::throw_ex<system_error>(errors::torrent_invalid_piece_layer);
+
+				auto& layer = piece_layers[t.root().to_string()].string();
+
+				for (auto const& h : piece_layer)
+					layer += h.to_string();
+			}
+		}
+		else if (atp.ti->v2())
+		{
+			// we must have piece layers for v2 torrents for them to be valid
+			// .torrent files
+			aux::throw_ex<system_error>(errors::torrent_missing_piece_layer);
+		}
+
+		// save web seeds
+		if (!atp.url_seeds.empty())
+		{
+			entry::list_type& url_list = ret["url-list"].list();
+			std::copy(atp.url_seeds.begin(), atp.url_seeds.end(), std::back_inserter(url_list));
+		}
+
+		if (!atp.http_seeds.empty())
+		{
+			entry::list_type& httpseeds_list = ret["httpseeds"].list();
+			std::copy(atp.http_seeds.begin(), atp.http_seeds.end(), std::back_inserter(httpseeds_list));
+		}
+
+		if (!atp.name.empty()) ret["name"] = atp.name;
```

Comment:
This looks suspicious. Is this field used in a torrent file?

Reply:
hm. no, probably not. "name" is pulled out of the info dictionary